### PR TITLE
Allow NamedPreparedStatement to prepare statement without retrieving auto generated columns

### LINKED ISSUE #26: Allow NamedPreparedStatement usage without retrieving auto generated columns
**Description:**
Usage of NamedPreparedStatement in NamedJdbcTemplate always creates a connection trying to retrieve an auto generated column with column name "ID". With the H2 DB upgrade (to 2.x), if column Name "ID" is not present, an exception is thrown.

`Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Column "ID" not found;`

### PR DESCRIPTION
**Pull request by vinumaddumage**
## Description
Issue #26 is related to how the [NamedPrepredStatement](https://github.com/wso2/carbon-utils/blob/master/components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java) constructors have been implemented and how they have been used in preparing sql queries in the [NamedJdbcTemplate](https://github.com/wso2/carbon-utils/blob/master/components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java), specifically with regards to the insert queries where if there is an auto generated field the corresponding field names should be passed to the connection (which will be used if they are requested to be fetched). Currently, NamedPreparedStatement always uses a default column name "ID " (which cannot be passed in the current implementation when executing a NamedJdbcTemplate method) as the returning column name even though it is may not be an insert query. The existing usage of NamedJdbcTemplate that could be found in IS (the Secret manager feature) always had the field "ID" in the table related to the query. If "ID" is not a field name the above exception is seen. This issue was not seen in previous releases which used H2 1.x since it just ignored it.

### CODE REVIEW

**Review by madurangasiriwardena — APPROVED** [review]

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java [hunk 1]
```diff
@@ -62,7 +62,30 @@ public class NamedPreparedStatement implements PreparedStatement {
...
+
+            if (repetition.get(sqlQuery.substring(pos + 1, end)) != null) {
+                for (int i = 0; i < repetition.get(sqlQuery.substring(pos + 1, end)) - 1; i++) {
+                    builder.append(", ?");
+                }
+            }
+
+            sqlQuery = String.format("%s %s %s", sqlQuery.substring(0, pos), builder.toString(),
+                    sqlQuery.substring(end + 1));
+        }
```
Comment (dewniMW) [inline components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java:86]: lets extract the duplicate code to a private method

Comment (dewniMW) [inline components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java]: let's extract the duplicate code to a private method

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java [hunk 2]
```diff
@@ -227,6 +202,43 @@ private List<Integer> getIndexList(String name) {
...
+
+        if (returningColumnName == null) {
+            preparedStatement = connection.prepareStatement(sqlQuery);
+        } else {
+            preparedStatement = connection.prepareStatement(sqlQuery, new String[]
+                    {JdbcUtils.getConvertedAutoGeneratedColumnName(connection.getMetaData()
+                            .getDatabaseProductName(), returningColumnName)});
+        }
+        return sqlQuery;
```
Comment (dewniMW) [inline components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java]: is this return value actually used?

**Review by dewniMW — APPROVED** [review]

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java [hunk 3]
```diff
@@ -258,36 +259,58 @@ public int executeUpdate(String query) throws DataAccessException {
...
-                        resultId = generatedKeys.getInt(1);
-                        logDebug("Newly inserted ID (Auto Increment ID) is {} for the bean {} ", resultId,
-                                bean);
-                    } else {
-                        throw new SQLException(JdbcConstants.ErrorCodes.ERROR_CODE_AUTO_GENERATED_ID_FAILURE.
-                                getErrorMessage());
+                try (NamedPreparedStatement namedPreparedStatement = new NamedPreparedStatement(connection, query,
+                        autoGenerateIdColumnName)) {
+                    doInternalUpdate(namedQueryFilter, namedPreparedStatement);
+                    logDebug("Mapping generated key (Auto Increment ID) to the object");
```
Comment (sahandilshan) [inline components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java:294]: Add a full stop at the end of the log message

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java [hunk 4]
```diff
@@ -258,36 +259,58 @@ public int executeUpdate(String query) throws DataAccessException {
...
+                    try (ResultSet generatedKeys = namedPreparedStatement.getGeneratedKeys()) {
+                        if (generatedKeys.next()) {
+                            resultId = generatedKeys.getInt(1);
+                            logDebug("Newly inserted ID (Auto Increment ID) is {} for the bean {} ", resultId,
```
Comment (sahandilshan) [inline components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java:298]: Full stop at the end of the log message

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java (hunk 1 above)
Comment (vinumaddumage) [reply components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java]: addressed by https://github.com/wso2/carbon-utils/pull/27/commits/ea0619889b213b68f87d29db92a404c3325b3935

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java (hunk 4 above)
Comment (vinumaddumage) [reply components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java:298]: updated 81e310b1cff1b6d5a106bd6ae81fd5137d85bcdf

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java (hunk 3 above)
Comment (vinumaddumage) [reply components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java:294]: updated 81e310b1cff1b6d5a106bd6ae81fd5137d85bcdf

File: components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java (hunk 2 above)
Comment (vinumaddumage) [reply components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedPreparedStatement.java]: Used to update the class attributes.

**Pull request merged**